### ISSUE TITLE
anchor scale

### ISSUE THREAD

**Issue by leap4133** [zylo117/Yet-Another-EfficientDet-Pytorch#171]
The code use the following equation to set the base anchor size, the size of the anchor with 1 and 1 the scale and ratio. 

base_anchor_size = self.anchor_scale * stride * scale
where 
self.anchor_scale = 4 (by default)
stride = [2 ** x for x in [3, 4, 5, 6, 7]]
scale = 1 for base anchor size

By the above equation, in P3, we got the base anchor size as 
4*(2**3)*1 = 32.

Is that too big for P3? I originally through it should be 8 by the way if I understand the paper correctly.

**Comment by zylo117** [comment]
It is 32,  anchor size 32 doesn't mean it's 32 on P3's feature map, but the corresponding size on the input images.

**Comment by leap4133** [comment]
@zylo117 I understand the size 32 is a corresponding size on the input image. But it should be 8, right? Since P3 down samples the image by a ratio of 8 (2**3). Why we manually enlarge the anchor by a constant of 4 (self.anchor_scale = 4)?

Consider the extreme case if we go for P7 feature map (spatial size is 4*4). Each "pixel" in such feature map is mapped to an 128*128 region on the input image without the scale of 4. If we enlarge it by 4. The anchor size will be a 512*512 for each of the 16 anchors. Does it make sense?

**Comment by zylo117** [comment]
Why not? Otherwise, how to locate some large objects if max anchor size is merely 128?

You can try running with anchor scale at 1. The boxes are too small to cover most of the objects.
![img_inferred_d0_this_repo_0](https://user-images.githubusercontent.com/5842006/79934752-e5d9e280-8485-11ea-9cfd-b1f8dfdf9abe.jpg)

However if you switch to 4. Then most of the objects can be bounded.
![img_inferred_d0_this_repo_0+a](https://user-images.githubusercontent.com/5842006/79934799-04d87480-8486-11ea-87eb-38af3d88ee83.jpg)

**Comment by leap4133** [comment]
I think if we remove the manual scaling. We need to re-train the regression part. Since the predicted boxes should be matched to the anchors in a different layer.

**Comment by zylo117** [comment]
Yes, but the largest anchor is still 128 with anchor scale 1. What if there is an object with size 256x256?

**Comment by jinfagang** [comment]
using 4 then tiny object missed out?

**Comment by leap4133** [comment]
using 4 means that we need to enlarge input image so that the small object can be detected.

**Comment by leap4133** [comment]
@zylo117 well the opposite side of the question is what if there is a small object such as 8*8. Since the min base anchor size is 32*32. Regression on dw, dh need to handle such extreme small case. A direct observation is that when I test the image on official google released tf checkpoint d0, I do not need to enlarge image even for a small car to be detected on an express way since their base anchor side is simply 1. For your D0 checkpoint, I have to enlarge to 1280 to has similar performance.

**Comment by jinfagang** [comment]
@leap4133 Mine is too. But coco also have massive tiny objects

**Comment by leap4133** [comment]
@jinfagang 
I would guess those tiny objects are also missed by the detector since we get low performance when we look at the mAP and recall for "small area" object in coco evaluation output. 

For small object, what I can think of is 
(1) apply smaller base-anchor size such as 1 and 2. In this way, the min anchor size is 8 and 16 respectively.
(2) use P2 feature map as an input to biFPN module.
(3) enlarge the image to ensure that the small object is within the detectable range of anchor scales which has min size of 32 currently.

Solution 1 means we need to finetune at least biFPN and cls/reg head
Solution 2 and 3 means we need more computation for inference and training

**Comment by Cli98** [comment]
> @jinfagang
> I would guess those tiny objects are also missed by the detector since we get low performance when we look at the mAP and recall for "small area" object in coco evaluation output.
> 
> For small object, what I can think of is
> (1) apply smaller base-anchor size such as 1 and 2. In this way, the min anchor size is 8 and 16 respectively.
> (2) use P2 feature map as an input to biFPN module.
> (3) enlarge the image to ensure that the small object is within the detectable range of anchor scales which has min size of 32 currently.
> 
> Solution 1 means we need to finetune at least biFPN and cls/reg head
> Solution 2 and 3 means we need more computation for inference and training

@leap4133 In case of your need to compute anchors, please try this repo,
https://github.com/Cli98/anchor_computation_tool

**Comment by kimalaacer** [comment]
> @jinfagang
> I would guess those tiny objects are also missed by the detector since we get low performance when we look at the mAP and recall for "small area" object in coco evaluation output.
> 
> For small object, what I can think of is
> (1) apply smaller base-anchor size such as 1 and 2. In this way, the min anchor size is 8 and 16 respectively.
> (2) use P2 feature map as an input to biFPN module.
> (3) enlarge the image to ensure that the small object is within the detectable range of anchor scales which has min size of 32 currently.
> 
> Solution 1 means we need to finetune at least biFPN and cls/reg head
> Solution 2 and 3 means we need more computation for inference and training

I am trying to apply solution 2 and 3 for small object detection? 
can you let me know where should I assign P2 as the input to BiFPN? 
have you tried it?
